### PR TITLE
Storybook 7

### PR DESCRIPTION
**Pull request by snowystinger**
Closes 

Attempting to upgrade us to storybook 7 so that we can develop against canary since sb7 changed their UI to use the new React render API and React has now removed the old version in canary.

Steps I followed:
run `npx storybook@7.6.17 upgrade`
move to new framework
dedupe storybook versions in the lockfile
deduped parcel
deduped babel

patched some files with @devongovett 's help

Storybook now runs!

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### CODE REVIEW

**Comment by sookmax** [comment]
> so that we can develop against canary since sb7 changed their UI to use the new React render API and React has now removed the old version in canary

Sorry I just casually stumbled upon this PR, and also I'm not sure what this comment 👆 really means 😂, but I think [Storybook 8](https://storybook.js.org/blog/storybook-8/) is out now? [npm versions](https://www.npmjs.com/package/storybook?activeTab=versions), and of course unless there's particular reason to choose v7, maybe upgrading to v8 would be better?

**Comment by reidbarber** [comment]
@sookmax Big emphasis on "attempting" . Upgrading majors for storybook isn't as easy as it sounds unfortunately. There usually ends up being a lot to change, and sometimes blockers. The migration assistants are really helpful, but with a large project and additional configurations, it's still a big task.

And to explain that comment, we currently run our tests in React 16, 17, 18, and [canary](https://react.dev/blog/2023/05/03/react-canaries). Running them with canary gives us a heads up on if something will break in a future release, so we can fix it earlier. In the latest canary, they've dropped support for the legacy ReactDOM.render API, so we need to switch everything to work with the newer root API, introduced in React 18. Upgrading storybook to the version that includes that will help.

**Comment by reidbarber** [comment]
I noticed that as well. I haven't tried running our storybook with canary, so not sure if any other issues arose when @snowystinger was debugging.

**Comment by snowystinger** [comment]
Hi @sookmax thanks for your interest. The changelog is referencing how storybook renders our components, and unfortunately not how storybook renders its UI/shell.
https://github.com/storybookjs/storybook/issues/18331#issuecomment-1162957861

**Comment by sookmax** [comment]
@snowystinger Thanks for taking time to explain this to me!

Forgive my ignorance, but the reason we care about what React render API the Storybook manager is using is because React as a dependency is being shared between React Aria and Storybook when testing in React 16, 17, 18, and canary? I somehow assumed Storybook would get its React separately and independently from that of React Aria, but I might be wrong. Sorry, It just a little confusing to me 😭.

> My hope in going to 7 is that 7-8 will be a smaller jump than 6-7.

This sounds reasonable!

> We also have some working repos with parcel + storybook on v7, so we have something to compare against.

Ah right, this definitely makes sense.

> storybook 8 is a bit more unknown and broke in interesting ways when I tried it.

Haha, I haven't tried v8 myself yet, but I can see that.. I contributed to Storybook 7 a little bit last year, and well.. their issues and PRs are not very well managed, and they were breaking things too much while moving forward 😅.

**Comment by sookmax** [comment]
> it's a peer dependency https://github.com/storybookjs/storybook/blob/next/code/frameworks/react-webpack5/package.json#L55 & https://github.com/storybookjs/storybook/blob/next/code/renderers/react/package.json#L89
for our parcel framework as well

Yeah you're right. A framework (bundler) and a renderer (react) specifying react and react-dom as peer dependencies make sense since they are responsible for rendering user's react code in [the preview (iframe) area](https://storybook.js.org/docs/addons#storybook-basics).

If I remember correctly (I might be wrong), In Storybook 7, the manager (Storybook UI) specifies react as its [devDependency](https://github.com/storybookjs/storybook/blob/next/code/ui/manager/package.json#L101-L102) and [pre-bundle itself using esbuild](https://github.com/storybookjs/storybook/blob/next/code/ui/manager/package.json#L71-L72) before distribution. In fact, I think the manager cannot specify react as a peer dependency because the framework and renderer combo might not even include react, yet the manager is always a React app.

Screenshot 2024-04-09 at 11 52 54 PM

I checked out `sb-7` branch and examined the installed `@storybook/manager` in `node_modules`, and it seems, in my case, one of the chunk files inside `dist/` contains react in its entirety:
Screenshot 2024-04-10 at 12 15 10 AM



But I think it wasn't necessarily the case with SB 6.5 now that I'm looking at `node_modules/@storybook/ui/package.json` in the main branch:

Screenshot 2024-04-10 at 12 03 36 AM

Anyways, I appreciate your work on the migration! It must be very challenging though.. haha.

P.S. It looks like the manager with react 18 was merged only into Storybook 8 [(release-8-0)](https://github.com/storybookjs/storybook/tree/release-8-0) by https://github.com/storybookjs/storybook/pull/24514

**Comment by LFDanLu** [comment]
some of these tests failures (random timeouts and jest worker being force exited) look pretty similar to the ones I've been getting in the treeview branch, perhaps meaning the babel is indeed to blame? Will try the upgrade on a branch from main

**Review by ktabors — COMMENTED** [review]
Thank you for doing this! I like the seperation of chromatic and chromatic-fc. The PR does a good job of hiding the difficulty of getting the correct versions and deduplicating.

**Comment by LFDanLu** [comment]
@snowystinger Still looking at the code but noticed that the storybook 17 and 16 builds are broken. The other builds (canary + 18) work just fine it seems though, controls/automated test story for Picker/themes/etc seem to be working well

**Comment by snowystinger** [comment]
> @snowystinger Still looking at the code but noticed that the storybook 17 and 16 builds are broken. The other builds (canary + 18) work just fine it seems though, controls/automated test story for Picker/themes/etc seem to be working well
> 
> Looks like it is trying to call `createRoot` which is only React 18+?

looks like problem with the shim, looking into it

**Review by reidbarber — DISMISSED** [review]

File: packages/@react-spectrum/checkbox/package.json [hunk 1]
```diff
@@ -41,7 +41,9 @@
     "@react-aria/interactions": "^3.21.1",
     "@react-spectrum/form": "^3.7.4",
     "@react-spectrum/label": "^3.16.4",
+    "@react-spectrum/layout": "^3.6.3",
     "@react-spectrum/utils": "^3.11.5",
+    "@react-spectrum/view": "^3.6.8",
```
Comment (LFDanLu) [inline packages/@react-spectrum/checkbox/package.json:46]: Where did these new deps come from? Seems to come up in a couple of the package.json changes, were they auto added by the IDE perhaps? The only imports for these are in the stories and stuff so we don't need them here right?

File: .eslintignore [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
 packages/@spectrum-icons/workflow/**
 packages/@spectrum-icons/illustrations/**
 packages/@spectrum-icons/express/**
 node_modules
 packages/*/*/dist
+packages/*/*/i18n
 packages/react-aria/dist
+packages/react-aria/i18n
 packages/react-aria-components/dist
+packages/react-aria-components/i18n
```
Comment (LFDanLu) [inline .eslintignore:13]: not sure this was intentional or not?

**Review by LFDanLu — COMMENTED** [review]
LGTM to me, storybooks seem to work and chromatic runs clean. Just some small changes that I assume were unintentional/from the merge/leftover from testing

File: .eslintignore (hunk 2 above)
Comment (snowystinger) [reply .eslintignore:13]: It was, they need to be added to our clean function as well.
I think they came from build, possibly in another branch. Either way, they need to be ignored and/or removed by clean_all

Comment (LFDanLu) [reply .eslintignore:13]: ah I misunderstood, I thought this was ignoring the i18n files like `useFilter`, makes sense

**Review by LFDanLu — APPROVED** [review]

**Review by reidbarber — APPROVED** [review]
Nice!

**Pull request merged**